config/helpers.php

```diff
@@ -790,60 +790,87 @@ function tc(string $key, int $count)
+    switch ($step['unit']) {
+        case 'second':
+            $interval = 'PT' . $step['size'] . 'S';
+            $base     = date('Y-m-d H:i:00', $date);
+            break;
+        case 'minute':
+            $interval = 'PT' . $step['size'] . 'M';
+            $base     = date('Y-m-d H:00:00', $date);
+            break;
+        case 'hour':
+            $interval = 'PT' . $step['size'] . 'H';
+            $base     = date('Y-m-d 00:00:00', $date);
+            break;
+        case 'day':
+            $interval = 'P' . $step['size'] . 'D';
+            $base     = date('Y-m-01 00:00:00', $date);
+            break;
+        case 'month':
+            $interval = 'P' . $step['size'] . 'M';
+            $base     = date('Y-01-01 00:00:00', $date);
+            break;
+        case 'year':
+            $interval = 'P' . $step['size'] . 'Y';
+            $century  = floor(date('Y', $date) / 100) * 100;
+            $base     = date($century . '-01-01 00:00:00', $date);
+            break;
     }
 
-    $parts = [
-        'second' => date('s', $date),
-        'minute' => date('i', $date),
-        'hour'   => date('H', $date),
-        'day'    => date('d', $date),
-        'month'  => date('m', $date),
-        'year'   => date('Y', $date),
-    ];
+    // create period from reference date to maximum date
+    // with points at each step-sized interval
+    $interval = new \DateInterval($interval);
+    $min      = new \DateTimeImmutable($base);
+    $max      = (new \DateTimeImmutable())->setTimestamp($date)->add($interval);
+    $period   = new \DatePeriod($min, $interval, $max);
 
-    $current = $parts[$step['unit']];
-    $nearest = round($current / $step['size']) * $step['size'];
-    $parts[$step['unit']] = $nearest;
+    // get step size in seconds for comparison
+    $step = $min->add($interval)->getTimestamp() - $min->getTimestamp();
 
-    foreach ($parts as $part => $value) {
-        if ($part === $step['unit']) {
-            break;
-        }
+    // get all potential interval points as options and
+    // reverse their order to prefer ceiling to nearest step
+    $options = array_reverse(iterator_to_array($period));
 
-        $parts[$part] = 0;
+    // loop through each option and check if option is
+    // less than half a step size away from actual timestamp
+    // (which makes it the nearest step)
+    foreach ($options as $option) {
+        if (abs($date - ($option = $option->getTimestamp())) <= $step/2) {
+            return $option;
+        }
     }
 
-    $timestamp = strtotime(
-        $parts['year'] . '-' .
-        str_pad($parts['month'], 2, 0, STR_PAD_LEFT) . '-' .
-        str_pad($parts['day'], 2, 0, STR_PAD_LEFT) . ' ' .
-        str_pad($parts['hour'], 2, 0, STR_PAD_LEFT) . ':' .
-        str_pad($parts['minute'], 2, 0, STR_PAD_LEFT) . ':' .
-        str_pad($parts['second'], 2, 0, STR_PAD_LEFT)
-    );
-
-    // on error, convert `false` into `null`
-    return $timestamp ?? null;
+    return null;
```

Comment:
In what case does this line work? 🤯 😅

Reply:
This return statement is the fallback if the period `$options` are empty if I understand it correctly.

Reply:
I felt like there should be an ultimate return in case nothing before matched. But I agree that I with the current setup, I think this line can't ever really been reached. 🤔